crates/sui-core/src/authority_aggregator.rs

```diff
@@ -962,12 +1073,87 @@ where
             Err(state) => {
                 self.record_process_transaction_metrics(tx_digest, &state);
                 let state = Self::record_non_quorum_effects_maybe(tx_digest, state);
-                Err(QuorumSignTransactionError {
-                    total_stake: self.committee.total_votes,
-                    good_stake: state.good_stake(),
+                Err(self.handle_process_transaction_error(tx_digest, state))
+            }
+        }
+    }
+
+    // TODO make sure this is epoch-boundary-safe
+    fn handle_process_transaction_error(
+        &self,
+        original_tx_digest: &TransactionDigest,
+        state: ProcessTransactionState,
+    ) -> AggregatorProcessTransactionError {
+        let validity_threshold = self.committee.validity_threshold();
+
+        if state.retryable {
+            if !state.conflicting_tx_digests.is_empty() {
+                if let Some((conflicting_tx, validators, conflicting_stake)) =
+                    state.conflicting_tx_digest_with_most_stake()
+                {
+                    let good_stake = state.good_stake();
+                    if good_stake >= validity_threshold && conflicting_stake >= validity_threshold {
+                        warn!(
+                            ?conflicting_tx,
+                            ?original_tx_digest,
+                            original_tx_stake = good_stake,
+                            tx_stake = conflicting_stake,
+                            "Equivocation detected: {:?}",
+                            validators
+                        );
+                        self.metrics.total_equivocation_detected.inc();
+                        return AggregatorProcessTransactionError::FatalConflictingTransaction {
+                            errors: state.errors,
+                            conflicting_tx_digests: state.conflicting_tx_digests,
+                        };
+                    } else if good_stake >= validity_threshold {
+                        // We have >= f+1 good stake on the current transaction so we will retry that,
+                        // no point in retrying conflicting ones.
```

Comment:
Imagine 4 validators each with 1 stake. 1 validator blesses tx1, 1 validator blesses tx2, now we try tx3 which gets the remaining 2 validators' blesses. Then it will always retry and hit `RetryableConflictingTransaction` until QD exhausted all retry attempts. 

Now with fix of suggested check (the `total_conflicting_stake_from_all_conflicitng_transactions` thing above), this would become less severe but still problematic. There would be byzantine validators messing around with non retryable errors. The ultimate check here is:
```
good >= validity && total_conflicts + non_retryable < validity
```